### PR TITLE
cqn2sql: dont forget to wrap `xpr`in parens

### PR DESCRIPTION
**Pull request by patricebender**
```log
sqlite> SELECT  2023||8 * 2 - 0 as something FROM Person;
40476
```
vs.
```log
sqlite> SELECT  2023|| (8 * 2 - 0) as something FROM Person;
202316
```

--> the xpr needs to be wrapped in parentheses:

```json
{
  "func": "concat",
  "args": [
    { "val": 2023 },
    {
      "xpr": [
        { "val": 8 },
        "*",
        { "val": 2 },
        "-",
        { "val": 0 }
      ]
    }
  ],
  "as": "something"
}
```

### CODE REVIEW

File: db-service/lib/cql-functions.js [hunk 1]
```diff
@@ -51,7 +51,7 @@ const StandardFunctions = {
    * @param {string} x
    * @returns {string}
    */
-  count: x => `count(${x || '*'})`,
+  count: x => `count${(x?.xpr && x ) || (x && `(${x})`) || '(*)'}`,
```
Comment (danjoa) [inline db-service/lib/cql-functions.js:54]:
```suggestion
  count: x => !x ? 'count(*)' : x.xpr ? `count${x}` : `count(${x})`,
```

Comment (danjoa) [reply db-service/lib/cql-functions.js:54]: What is an example of such expressions in `count()`?

Comment (danjoa) [reply db-service/lib/cql-functions.js:54]: Looks like this is a follow up fix for always wrapping `xpr` in `expr()`? → likely not good

File: db-service/lib/cqn2sql.js [hunk 2]
```diff
@@ -664,7 +664,7 @@ class CQN2SQLRenderer {
     if ('param' in x) return wrap(this.param(x))
     if ('ref' in x) return wrap(this.ref(x))
     if ('val' in x) return wrap(this.val(x))
-    if ('xpr' in x) return wrap(this.xpr(x))
+    if ('xpr' in x) return wrap(`(${this.xpr(x)})`)
```
Comment (danjoa) [inline db-service/lib/cqn2sql.js:667]: Looks like far too much parenthesis

**Review by danjoa — CHANGES_REQUESTED** [review]
I don't understand, why exactly we need that.
And if so, there has to be a better way

**Comment by patricebender** [comment]
> I don't understand, why exactly we need that. And if so, there has to be a better way

https://github.com/cap-js/cds-dbs/issues/192

File: db-service/lib/cqn2sql.js (hunk 2 above)
Comment (BobdenOs) [reply db-service/lib/cqn2sql.js:667]: I double checked and this is indeed not the correct place to do this. It is already being done where it should be done [here](https://github.com/cap-js/cds-dbs/blob/main/db-service/lib/cqn2sql.js#L684).

While constructing a reproduction query I found another issue.
```cql
-- CQL input
SELECT (1.0+2.0)*(1.0/2.0);
SELECT (1+2)*(1/2); -- is equal in cqn

-- SQL expectation
SELECT (1.0 + 2.0) * (1.0 / 2.0); -- 1.5

-- SQL output
SELECT (1 + 2) * (1 / 2); -- 0
```

**Comment by danjoa** [comment]
The problem is the implementation of the concat() function for SQLite. 
The changes in PR are too broad for that.

**Comment by BobdenOs** [comment]
@danjoa the implementation was adjusted to only change the `concat` implementation.
[here](https://github.com/cap-js/cds-dbs/pull/196/files#diff-d9ef4ae327f7e5954027eae8acf2866e34a6a365a11e60a3dc5ff535a210c33fR42)

**Comment by danjoa** [comment]
> @danjoa the implementation was adjusted to only change the `concat` implementation.
> [here](https://github.com/cap-js/cds-dbs/pull/196/files#diff-d9ef4ae327f7e5954027eae8acf2866e34a6a365a11e60a3dc5ff535a210c33fR42)

Yes, this is the correct fix. 
```js
  concat: (...args) => args.map(x => x.xpr ? `(${x})` : x).join(' || '),
```

Should work as well I think. We don't need to wrap gals and refs into parentheses.

File: db-service/lib/cql-functions.js [hunk 3]
```diff
@@ -39,7 +39,8 @@ const StandardFunctions = {
    * @param  {...string} args
    * @returns {string}
    */
-  concat: (...args) => args.join('||'),
+  // concat: (...args) => args.map(a => `(${a})`).join('||'), // More brackets
+  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join('||'), // Possible errors?
```
Comment (patricebender) [inline db-service/lib/cql-functions.js:43]:
```suggestion
  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join('||'),
```

File: db-service/test/cqn2sql/expression.test.js [hunk 4]
```diff
@@ -27,6 +27,27 @@ describe('expressions', () => {
     const { sql, values } = cqn2sql(cqn)
     expect({ sql, values }).toMatchSnapshot()
   })
+  it.only('ref, String operator and value', () => {
```
Comment (patricebender) [inline db-service/test/cqn2sql/expression.test.js:30]:
```suggestion
  xtest('ref, String operator and value', () => {
```

File: db-service/lib/cql-functions.js [hunk 5]
```diff
@@ -39,7 +39,8 @@ const StandardFunctions = {
    * @param  {...string} args
    * @returns {string}
    */
-  concat: (...args) => args.join('||'),
+  // concat: (...args) => args.map(a => `(${a})`).join('||'), // More brackets
+  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join(' || '),
```
Comment (danjoa) [inline db-service/lib/cql-functions.js:43]:
```suggestion
  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join(' || '),
```

File: db-service/lib/cql-functions.js [hunk 6]
```diff
@@ -39,7 +39,7 @@ const StandardFunctions = {
    * @param  {...string} args
    * @returns {string}
    */
-  concat: (...args) => args.join('||'),
+  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join(' || '),
```
Comment (danjoa) [inline db-service/lib/cql-functions.js:42]:
```suggestion
  concat: (...args) => args.map(a => a.xpr ? `(${a})` : a).join(' || '),
  
```

**Review by danjoa — APPROVED** [review]

**Pull request merged**